src/Upload/StreamableUploader.php

```diff
@@ -0,0 +1,165 @@
+<?php
+/**
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may ob`tain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Upload;
+
+use GuzzleHttp\Psr7\BufferStream;
+use GuzzleHttp\Psr7\Request;
+
+/**
+ * Uploader that is a special case of the ResumableUploader where we can write
+ * the file contents in a streaming manner.
+ */
+class StreamableUploader extends ResumableUploader
+{
+    const DEFAULT_WRITE_CHUNK_SIZE = 262144;
+
+    /**
+     * @param RequestWrapper $requestWrapper
+     * @param string|resource|StreamInterface $data
+     * @param string $uri
+     * @param array $options [optional] {
+     *     Optional configuration.
+     *
+     *     @type array $metadata Metadata on the resource.
+     *     @type int $chunkSize Size of the chunks to send incrementally during
+     *           a resumable upload. Must be in multiples of 262144 bytes.
+     *     @type array $httpOptions HTTP client specific configuration options.
+     *     @type int $retries Number of retries for a failed request.
+     *           **Defaults to** `3`.
+     *     @type string $contentType Content type of the resource.
+     * }
+     */
+    public function __construct()
+    {
+        call_user_func_array(array($this, 'parent::__construct'), func_get_args());
```

Comment:
We lose out on type hinting with this approach, which is nice for IDEs that might read that information.